text/0000-router-service.md

```diff
@@ -0,0 +1,440 @@
+
+```js
+router.currentRoute.find(info => info.name === 'people').params
+```
+
+## RouteInfoWithAttributes
+
+This type is almost identical to `RouteInfo`, except it has one additional property named `attributes`. The attributes contain the data that was loaded for this route, which is typically just `{ model }`.
+
+## Transition Object
+
+A `transition` argument is passed to `Route#beforeModel`, `Route#model`, `Route#afterModel`, `Route#willTransition`, and `Router#willTransition`. Today `transition`'s public API is only really `abort()` and `retry()`.
+
+### New Properties: `from` and `to`
+
+I'm proposing we add `from` and `to` properties on `transition` whose values are `RouteInfo` instances representing the initial and final leafmost routes for this transition. Like all RouteInfos, these are read-only and internally immutable. They are not observable, because a  `transition` instance is never changed after creation.
+
+On an initial full-page load, the `from` property will be `null`. This creates a public API for distinguishing in-app transitions from full-page reloads.
+
+### Example: testing whether route will remain active
+
+Here's an example showing how `willTransition` can figure out if the current route will remain active after the transition:
+
+```js
+willTransition(transition) {
+  if (!this.transition.to.find(route => route.name === this.routeName)) {
+    alert("Please save or cancel your changes.");
+    transition.abort();
+  }
+}
+```
+
+### Example: parent redirecting to a fallback model
+
+Here's an example of a parent route that can redirect to a fallback model, without losing its child route:
+
+```js
+this.route('person', { path: '/person/:person_id' }, function() {
+  this.route('index');
+  this.route('detail');
+});
+
+//PersonRoute
+const fallbackPersonId = 0;
+model({ personId }, transition) {
+  return this.get('store').find('person', personId).catch(err => {
+    this.replaceWith(transition.to.name, fallbackPersonId);
+  });
+}
+
+// If personId 5 is invalid, and the user visits /person/5/detail, they will get
+// redirected to /person/0/detail. And /person/5 will get redirected to /person/0.
+```
+
+
+### Actively discourage use of private API
+
+This RFC provides public API for doing the things people have become accustomed to doing via private API. To eliminate confusion over the correct way, we should hide all the private API away behind symbols, and provide deprecation warnings per our usual release policy around breaking "widely-used private APIs".
+
+Some of the private APIs we should mark and warn include:
+
+ - transition.state
+ - transition.params
+ - `lookup('router:main')` (should use `service:router` instead)
+
+
+## Dynamically-Scoped Variables
```

Comment:
You don't need a getter-with-side-effects. You can have a component that takes `routeInfo` as an input parameter, and in `didRender` check to see whether the latest `routeInfo` is actually different enough to warrant the action.

(Doing a stability check is necessary because your component would see a new value even when only one of its child routes was rerendering, because `routeInfo` contains a comprehensive picture of the whole route hierarchy.)